### PR TITLE
added useful error messages why a subcourse can not be promoted

### PR DESCRIPTION
**Pull request by paul-jaworski**

### CODE REVIEW

File: common/courses/notifications.ts [hunk 1]
```diff
@@ -185,8 +185,47 @@ export const canPromoteSubcourse = async (subcourse: Prisma.subcourse, attempted
 export async function sendPupilCoursePromotion(subcourse: Prisma.subcourse, promotionType: Prisma.subcourse_promotion_type_enum) {
     const { allowed, reason } = await canPromoteSubcourse(subcourse, promotionType);
     if (!allowed) {
-        logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
```
Comment (JeangelLF) [inline common/courses/notifications.ts:188]: I think we could leave this here and remove all the duplicates in each `if` statement :)

File: common/courses/notifications.ts [hunk 2]
```diff
@@ -185,8 +185,47 @@ export const canPromoteSubcourse = async (subcourse: Prisma.subcourse, attempted
...
+            throw new Error(`Instructors can only promote a Subcourse once!`);
+        }
+        if (reason == 'no-admin-promotions-left') {
+            logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
+            throw new Error(`Admins can only promote a Subcourse once!`);
+        }
+        if (reason == 'recently-promoted') {
+            logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
+            throw new Error(`Subcourse(${subcourse.id}) has already been promoted less than 3 days ago!`);
+        }
```
Comment (JeangelLF) [inline common/courses/notifications.ts:228]: Just to be completely sure that this fails, even if `reason` doesn't match any of the `if`s we should throw an error in the end
```
        throw new Error(`Promotion for Subcourse(${subcourse.id}) is not valid!`);
```

File: common/courses/notifications.ts [hunk 3]
```diff
@@ -185,8 +185,47 @@ export const canPromoteSubcourse = async (subcourse: Prisma.subcourse, attempted
 export async function sendPupilCoursePromotion(subcourse: Prisma.subcourse, promotionType: Prisma.subcourse_promotion_type_enum) {
     const { allowed, reason } = await canPromoteSubcourse(subcourse, promotionType);
     if (!allowed) {
-        logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
-        throw new Error(`Promotion for Subcourse(${subcourse.id}) is not valid!`);
+        if (reason == 'invalid-promotion-type') {
+            logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
+            throw new Error(`Promotion type for Subcourse(${subcourse.id}) is not valid!`);
```
Comment (Jonasdoubleyou) [inline common/courses/notifications.ts:190]: What is the purpose of this change? Throwing an error will be reported as an internal error to the client.

**Comment by paul-jaworski** [comment]
This PR belongs to the Issue: https://github.com/corona-school/project-user/issues/1197

How should I throw the Error instead?

File: common/courses/notifications.ts (hunk 3 above)
Comment (paul-jaworski) [reply common/courses/notifications.ts:190]: This PR belongs to the Issue: corona-school/project-user#1197

I searched the code and found this way of throwing Errors. How should I throw the Error instead?

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:190]: Extend https://github.com/corona-school/backend/blob/master/common/util/error.ts with a new error class, maybe "NotAllowedError"

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:190]: That error class should have the reason as property

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:190]: Then adapt formatMessage in https://github.com/corona-school/backend/blob/master/graphql/error.ts to pass that reason to the client. It can then be catched and handled in the frontend.

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:190]: Maybe also adapt https://github.com/corona-school/backend/blob/master/common/util/decision.ts assertAllowed to throw that new error class and reuse that here

Comment (paul-jaworski) [reply common/courses/notifications.ts:190]: I created a new commit to implement your suggestion. However this seems like duplicate code to me. Maybe it would make more sense to export a new client error variable called "NotAllowedError" and then give the reason as the publicMessage?

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:190]: Ja well "That error class should have the reason as property" then it would not be redundant. 

But would also be OK to just create a new error type and then set the message to the reason

Comment (Jonasdoubleyou) [reply common/courses/notifications.ts:191]: throw new NotAllowedError(reason); ?

Comment (paul-jaworski) [reply common/courses/notifications.ts]: Yes I looked at how the other types of client errors were thrown and imitated that.

File: common/courses/notifications.ts [hunk 4]
```diff
@@ -186,7 +187,29 @@ export async function sendPupilCoursePromotion(subcourse: Prisma.subcourse, prom
...
+        } else if (reason == 'already-auto-promoted') {
+            throw new NotAllowedError(`Subcourse(${subcourse.id}) is already promoted automatically!`);
+        } else if (reason == 'no-instructor-promotions-left') {
+            throw new NotAllowedError(`Instructors can only promote a Subcourse once!`);
+        } else if (reason == 'no-admin-promotions-left') {
+            throw new NotAllowedError(`Admins can only promote a Subcourse once!`);
+        } else if (reason == 'recently-promoted') {
+            throw new NotAllowedError(`Subcourse(${subcourse.id}) has already been promoted less than 3 days ago!`);
+        } else {
+            throw new NotAllowedError('Subcourse(${subcourse.id}) can not be promoted for a different reason: ${reason}!');
```
Comment (JeangelLF) [inline common/courses/notifications.ts:211]: This should use ` instead of '

**Review by JeangelLF — CHANGES_REQUESTED** [review]

File: common/courses/notifications.ts (hunk 3 above)
Comment (Jonasdoubleyou) [reply common/courses/notifications.ts]: Again: throw new NotAllowedError(reason); - there is no point in showing english errors to our users, translation to a text should be done in the frontend.

File: common/courses/notifications.ts [hunk 5]
```diff
@@ -185,8 +186,7 @@ export const canPromoteSubcourse = async (subcourse: Prisma.subcourse, attempted
 export async function sendPupilCoursePromotion(subcourse: Prisma.subcourse, promotionType: Prisma.subcourse_promotion_type_enum) {
     const { allowed, reason } = await canPromoteSubcourse(subcourse, promotionType);
     if (!allowed) {
-        logger.info(`Can't promote Subcourse(${subcourse.id}). Reason: ${reason}`);
```
Comment (JeangelLF) [inline common/courses/notifications.ts:188]: Could we keep the log? It's very helpful for debugging purposes :)

**Review by JeangelLF — APPROVED** [review]
Thank you!
